core/controller/cluster/controller_basic.go

```diff
@@ -681,13 +691,26 @@ func (c *controller) UpdateCluster(ctx context.Context, clusterID uint,
 	}
 
 	// 5. update cluster in db
-	clusterModel := r.toClusterModel(cluster, templateRelease, er)
 	// todo: atomicity
 	cluster, err = c.clusterMgr.UpdateByID(ctx, clusterID, clusterModel)
 	if err != nil {
 		return nil, err
 	}
 
+	// 7. update cluster tags
+	tagsInDB, err := c.tagMgr.ListByResourceTypeID(ctx, common.ResourceCluster, clusterID)
+	if err != nil {
+		return nil, err
+	}
+	if !tagmodels.Tags(tags).Eq(tagsInDB) {
```

Comment:
It could be wrong if `tags` is part of `tagsInDB` when the user would like to remove some tags.